File: app/src/main/java/com/lucagiorgetti/surprix/utility/dao/ProducerDao.java
package com.lucagiorgetti.surprix.utility.dao;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.lucagiorgetti.surprix.SurprixApplication;
import com.lucagiorgetti.surprix.listenerInterfaces.CallbackInterface;
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback;
import com.lucagiorgetti.surprix.model.Producer;
import com.lucagiorgetti.surprix.model.Year;

import java.util.ArrayList;
import java.util.List;

public class ProducerDao {
    private static DatabaseReference producers = SurprixApplication.getInstance().getDatabaseReference().child("producers");
    private static YearDao yearDao = new YearDao();

    public static void getProducerYears(String producerId, final FirebaseListCallback<Year> listen) {
        listen.onStart();

        final ArrayList<Year> years = new ArrayList<>();

        producers.child(producerId).child("years").orderByChild("year").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (DataSnapshot d : dataSnapshot.getChildren()) {
                        yearDao.getYearById(d.getKey(), new CallbackInterface<Year>() {
                            @Override
                            public void onStart() {

                            }

                            @Override
                            public void onSuccess(Year item) {
                                years.add(item);
                                listen.onSuccess(years);
                            }

                            @Override
                            public void onFailure() {

                            }
                        });
                    }
                } else {
                    listen.onSuccess(null);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
            }
        });
    }

    public static void getProducers(final FirebaseListCallback<Producer> listen) {
        listen.onStart();

        final List<Producer> prods = new ArrayList<>();

        producers.orderByChild("order").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (DataSnapshot d : dataSnapshot.getChildren()) {
                        Producer p = d.getValue(Producer.class);
                        prods.add(p);
                        listen.onSuccess(prods);
                    }
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                listen.onFailure();
            }
        });
    }

    public static void getProducerById(String producerId, CallbackInterface<Producer> listen){
        producers.child(producerId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    Producer producer = snapshot.getValue(Producer.class);
                    listen.onSuccess(producer);
                }
                listen.onFailure();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }
}

File: app/src/main/java/com/lucagiorgetti/surprix/utility/dao/ProducerDao.kt
package com.lucagiorgetti.surprix.utility.dao

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.lucagiorgetti.surprix.SurprixApplication.Companion.getInstance
import com.lucagiorgetti.surprix.listenerInterfaces.CallbackInterface
import com.lucagiorgetti.surprix.listenerInterfaces.FirebaseListCallback
import com.lucagiorgetti.surprix.model.Producer
import com.lucagiorgetti.surprix.model.Year

object ProducerDao {
    private val producers = getInstance().databaseReference!!.child("producers")
    fun getProducerYears(producerId: String?, listen: FirebaseListCallback<Year>) {
        listen.onStart()
        val years = ArrayList<Year>()
        producers.child(producerId!!).child("years").orderByChild("year").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (d in dataSnapshot.children) {
                        YearDao.getYearById(d.key, object : CallbackInterface<Year> {
                            override fun onStart() {}
                            override fun onSuccess(item: Year) {
                                years.add(item)
                                listen.onSuccess(years)
                            }

                            override fun onFailure() {}
                        })
                    }
                } else {
                    listen.onSuccess(years)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun getProducers(listen: FirebaseListCallback<Producer>) {
        listen.onStart()
        val prods: MutableList<Producer> = ArrayList()
        producers.orderByChild("order").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    for (d in dataSnapshot.children) {
                        val p = d.getValue(Producer::class.java)!!
                        prods.add(p)
                        listen.onSuccess(prods)
                    }
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
                listen.onFailure()
            }
        })
    }

    fun getProducerById(producerId: String?, listen: CallbackInterface<Producer>) {
        producers.child(producerId!!).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    val producer = snapshot.getValue(Producer::class.java)!!
                    listen.onSuccess(producer)
                }
                listen.onFailure()
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }
}
